File: problems/424/compute/wave5/C116_endpoint_invariant.cpp
#include <algorithm>
#include <bit>
#include <chrono>
#include <cstdint>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <limits>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

constexpr std::uint32_t kMaximumLimit = 4'000'000'000U;
constexpr std::uint64_t kFnvOffset = 14'695'981'039'346'656'037ULL;
constexpr std::uint64_t kFnvPrime = 1'099'511'628'211ULL;

enum class State : std::uint8_t {
    other = 0,
    generated = 1,
    splitless = 2,
    hard = 3,
};

bool allowed(std::uint32_t n) {
    return n >= 2 && n % 3 != 1;
}

void require(bool condition, const char* message) {
    if (!condition) throw std::runtime_error(message);
}

void fnv_byte(std::uint64_t& digest, std::uint8_t value) {
    digest ^= value;
    digest *= kFnvPrime;
}

void fnv_u64(std::uint64_t& digest, std::uint64_t value) {
    for (unsigned shift = 0; shift < 64; shift += 8) {
        fnv_byte(digest, static_cast<std::uint8_t>(value >> shift));
    }
}

std::string hex_u64(std::uint64_t value) {
    std::ostringstream out;
    out << std::hex << std::setfill('0') << std::setw(16) << value;
    return out.str();
}

void build_odd_spf(std::uint32_t maximum, std::vector<std::uint16_t>& odd_spf) {
    for (std::uint32_t p = 3; static_cast<std::uint64_t>(p) * p <= maximum; p += 2) {
        if (odd_spf[p >> 1] != 0) continue;
        const std::uint64_t step = 2ULL * p;
        for (std::uint64_t multiple = static_cast<std::uint64_t>(p) * p;
             multiple <= maximum;
             multiple += step) {
            auto& entry = odd_spf[static_cast<std::size_t>(multiple >> 1)];
            if (entry == 0) entry = static_cast<std::uint16_t>(p);
        }
    }
}

std::uint32_t prime_factor(
    std::uint32_t value,
    const std::vector<std::uint16_t>& odd_spf
) {
    if ((value & 1U) == 0) return 2;
    const auto factor = odd_spf[value >> 1];
    return factor == 0 ? value : factor;
}

void enumerate_divisors(
    std::uint32_t value,
    const std::vector<std::uint16_t>& odd_spf,
    std::vector<std::uint32_t>& divisors
) {
    divisors.clear();
    divisors.push_back(1);
    auto remaining = value;
    while (remaining > 1) {
        const auto p = prime_factor(remaining, odd_spf);
        const auto old_size = divisors.size();
        std::uint32_t power = 1;
        do {
            remaining /= p;
            power *= p;
            for (std::size_t i = 0; i < old_size; ++i) {
                divisors.push_back(divisors[i] * power);
            }
        } while (remaining > 1 && remaining % p == 0);
    }
}

State classify(
    std::uint32_t n,
    const std::vector<std::uint32_t>& divisors,
    const std::vector<State>& state
) {
    const auto product = n + 1;
    bool has_pair = false;
    for (const auto left : divisors) {
        if (left < 2) continue;
        const auto right = product / left;
        if (left >= right || !allowed(left) || !allowed(right)) continue;
        has_pair = true;
        if (state[left] == State::generated && state[right] == State::generated) {
            return State::generated;
        }
    }
    if (!has_pair) return State::splitless;
    if ((n & 1U) == 0) {
        if (product % 3 != 0) return State::hard;
        const auto parent = product / 3;
        if (!allowed(parent) || parent == 3) return State::hard;
    }
    return State::other;
}

std::uint32_t seed_root(std::uint32_t endpoint) {
    require(endpoint >= 3 && (endpoint & 1U) != 0, "seed root requires odd endpoint");
    return ((endpoint - 1) >> std::countr_zero(endpoint - 1)) + 1;
}

const char* state_name(State state) {
    switch (state) {
        case State::other: return "other_hole";
        case State::generated: return "generated";
        case State::splitless: return "structural_splitless";
        case State::hard: return "hard";
    }
    throw std::runtime_error("unknown state");
}

struct Counts {
    std::uint32_t d = 0;
    std::uint32_t s = 0;
    std::uint32_t t = 0;
    std::uint32_t e0 = 0;
    std::uint32_t e1 = 0;
    std::int32_t minimum_prefix_slack = std::numeric_limits<std::int32_t>::max();
};

Counts endpoint_counts(
    std::uint32_t n,
    const std::vector<std::uint32_t>& divisors,
    const std::vector<State>& state,
    std::vector<std::uint32_t>& pair_lefts
) {
    Counts counts;
    const auto product = n + 1;
    pair_lefts.clear();
    for (const auto left : divisors) {
        if (left < 2) continue;
        const auto right = product / left;
        if (left >= right || !allowed(left) || !allowed(right)) continue;
        pair_lefts.push_back(left);
    }
    std::sort(pair_lefts.begin(), pair_lefts.end());
    std::uint32_t prefix_t = 0;
    for (const auto left : pair_lefts) {
        const auto right = product / left;
        ++counts.d;
        bool structural_pair = false;
        std::uint32_t missing = 0;
        for (const auto endpoint : {left, right}) {
            if (state[endpoint] == State::generated) continue;
            ++missing;
            const auto root = seed_root(endpoint);
            require(root < endpoint && allowed(root), "invalid seed root");
            require(state[root] != State::generated, "missing endpoint has generated root");
            if (state[root] == State::splitless) {
                ++counts.e0;
                structural_pair = true;
            } else {
                ++counts.e1;
            }
        }
        require(missing >= 1, "unblocked hard pair");
        counts.s += static_cast<std::uint32_t>(structural_pair);
        const auto canonical = state[left] != State::generated ? left : right;
        const auto canonical_structural = static_cast<std::uint32_t>(
            state[seed_root(canonical)] == State::splitless
        );
        counts.t += canonical_structural;
        prefix_t += canonical_structural;
        const auto prefix_slack = static_cast<std::int32_t>(2 * prefix_t)
            - static_cast<std::int32_t>(counts.d) + 8;
        counts.minimum_prefix_slack = std::min(
            counts.minimum_prefix_slack,
            prefix_slack
        );
    }
    return counts;
}

std::string describe_hard(
    std::uint32_t n,
    const std::vector<std::uint32_t>& divisors,
    const std::vector<State>& state,
    Counts counts
) {
    std::ostringstream out;
    const auto product = n + 1;
    out << "{\"h\":" << n << ",\"product\":" << product
        << ",\"d\":" << counts.d << ",\"s\":" << counts.s
        << ",\"t\":" << counts.t
        << ",\"E0_count\":" << counts.e0 << ",\"E1_count\":" << counts.e1
        << ",\"endpoint_imbalance_slack\":"
        << static_cast<std::int64_t>(counts.e0) + 8 - counts.e1
        << ",\"power_bridge_slack\":"
        << 3 * static_cast<std::int64_t>(counts.s) - counts.d + 8
        << ",\"canonical_power_slack\":"
        << 2 * static_cast<std::int64_t>(counts.t) - counts.d + 8
        << ",\"minimum_canonical_prefix_slack\":"
        << counts.minimum_prefix_slack
        << ",\"pairs\":[";
    std::vector<std::uint32_t> pair_lefts;
    for (const auto left : divisors) {
        if (left < 2) continue;
        const auto right = product / left;
        if (left < right && allowed(left) && allowed(right)) pair_lefts.push_back(left);
    }
    std::sort(pair_lefts.begin(), pair_lefts.end());
    bool first_pair = true;
    std::uint32_t prefix_index = 0;
    std::uint32_t prefix_t = 0;
    for (const auto left : pair_lefts) {
        const auto right = product / left;
        if (!first_pair) out << ',';
        first_pair = false;
        ++prefix_index;
        bool structural_pair = false;
        out << "{\"prefix_index\":" << prefix_index
            << ",\"pair\":[" << left << ',' << right << "],\"endpoints\":[";
        for (unsigned i = 0; i < 2; ++i) {
            if (i != 0) out << ',';
            const auto endpoint = i == 0 ? left : right;
            out << "{\"value\":" << endpoint << ",\"state\":\""
                << state_name(state[endpoint]) << "\",\"root\":";
            if (state[endpoint] == State::generated) {
                out << "null,\"root_state\":null}";
            } else {
                const auto root = seed_root(endpoint);
                structural_pair = structural_pair || state[root] == State::splitless;
                out << root << ",\"root_state\":\"" << state_name(state[root]) << "\"}";
            }
        }
        const auto canonical = state[left] != State::generated ? left : right;
        const auto canonical_root = seed_root(canonical);
        prefix_t += static_cast<std::uint32_t>(
            state[canonical_root] == State::splitless
        );
        out << "],\"counted_in_s\":" << (structural_pair ? "true" : "false")
            << ",\"canonical_blocker\":" << canonical
            << ",\"canonical_root\":" << canonical_root
            << ",\"counted_in_t\":"
            << (state[canonical_root] == State::splitless ? "true" : "false")
            << ",\"prefix_t\":" << prefix_t
            << ",\"prefix_slack\":"
            << 2 * static_cast<std::int64_t>(prefix_t) - prefix_index + 8 << '}';
    }
    out << "]}";
    return out.str();
}

void write_optional(std::ostream& out, const std::string& value) {
    if (value.empty()) out << "null"; else out << value;
}

}  // namespace

int main(int argc, char** argv) {
    try {
        if (argc != 3) {
            std::cerr << "usage: C116_endpoint_invariant LIMIT OUTPUT_JSON\n";
            return 2;
        }
        const auto parsed_limit = std::stoull(argv[1]);
        if (parsed_limit < 534 || parsed_limit > kMaximumLimit) {
            throw std::runtime_error("LIMIT must lie in [534,4000000000]");
        }
        const auto limit = static_cast<std::uint32_t>(parsed_limit);
        const auto started = std::chrono::steady_clock::now();
        std::vector<std::uint16_t> odd_spf(
            static_cast<std::size_t>((limit + 1) / 2) + 1
        );
        build_odd_spf(limit + 1, odd_spf);
        std::vector<State> state(static_cast<std::size_t>(limit) + 1);
        std::vector<std::uint32_t> divisors;
        divisors.reserve(4096);
        std::vector<std::uint32_t> pair_lefts;
        pair_lefts.reserve(256);

        std::uint64_t hard_count = 0;
        std::uint32_t maximum_d = 0;
        std::uint64_t classification_digest = kFnvOffset;
        std::uint64_t metric_digest = kFnvOffset;
        std::int64_t minimum_endpoint_slack = std::numeric_limits<std::int64_t>::max();
        std::int64_t minimum_power_slack = std::numeric_limits<std::int64_t>::max();
        std::int64_t minimum_canonical_slack = std::numeric_limits<std::int64_t>::max();
        std::int64_t minimum_canonical_prefix_slack =
            std::numeric_limits<std::int64_t>::max();
        std::string minimum_endpoint_detail;
        std::string minimum_power_detail;
        std::string minimum_canonical_detail;
        std::string minimum_canonical_prefix_detail;
        std::string endpoint_failure;
        std::string power_failure;
        std::string canonical_failure;
        std::string canonical_prefix_failure;

        for (std::uint32_t n = 2; n <= limit; ++n) {
            State current = State::other;
            if (n == 2 || n == 3) {
                current = State::generated;
            } else if (allowed(n)) {
                enumerate_divisors(n + 1, odd_spf, divisors);
                current = classify(n, divisors, state);
            }
            state[n] = current;
            fnv_byte(classification_digest, static_cast<std::uint8_t>(current));
            if (current != State::hard) continue;

            ++hard_count;
            const auto counts = endpoint_counts(n, divisors, state, pair_lefts);
            require(counts.d > 0 && counts.s <= counts.d, "invalid hard counts");
            require(counts.d - counts.s <= counts.e1, "first counting bridge failed");
            require(counts.e0 <= 2 * counts.s, "second counting bridge failed");
            require(counts.t <= counts.s, "canonical count exceeds structural pairs");
            maximum_d = std::max(maximum_d, counts.d);
            fnv_u64(metric_digest, n);
            fnv_u64(metric_digest, counts.d);
            fnv_u64(metric_digest, counts.s);

            const auto endpoint_slack =
                static_cast<std::int64_t>(counts.e0) + 8 - counts.e1;
            const auto power_slack =
                3 * static_cast<std::int64_t>(counts.s) - counts.d + 8;
            const auto canonical_slack =
                2 * static_cast<std::int64_t>(counts.t) - counts.d + 8;
            if (endpoint_slack < minimum_endpoint_slack) {
                minimum_endpoint_slack = endpoint_slack;
                minimum_endpoint_detail = describe_hard(n, divisors, state, counts);
            }
            if (power_slack < minimum_power_slack) {
                minimum_power_slack = power_slack;
                minimum_power_detail = describe_hard(n, divisors, state, counts);
            }
            if (canonical_slack < minimum_canonical_slack) {
                minimum_canonical_slack = canonical_slack;
                minimum_canonical_detail = describe_hard(n, divisors, state, counts);
            }
            if (counts.minimum_prefix_slack < minimum_canonical_prefix_slack) {
                minimum_canonical_prefix_slack = counts.minimum_prefix_slack;
                minimum_canonical_prefix_detail = describe_hard(n, divisors, state, counts);
            }
            if (endpoint_slack < 0 && endpoint_failure.empty()) {
                endpoint_failure = describe_hard(n, divisors, state, counts);
            }
            if (power_slack < 0 && power_failure.empty()) {
                power_failure = describe_hard(n, divisors, state, counts);
            }
            if (canonical_slack < 0 && canonical_failure.empty()) {
                canonical_failure = describe_hard(n, divisors, state, counts);
            }
            if (counts.minimum_prefix_slack < 0 && canonical_prefix_failure.empty()) {
                canonical_prefix_failure = describe_hard(n, divisors, state, counts);
            }
        }

        std::ofstream out(argv[2], std::ios::binary);
        if (!out) throw std::runtime_error("could not open output JSON");
        out << "{\n"
            << "  \"schema\":\"C116-endpoint-invariant-cpp-v2\",\n"
            << "  \"limit\":" << limit << ",\n"
            << "  \"exact_integer_acceptance\":true,\n"
            << "  \"hard_holes\":" << hard_count << ",\n"
            << "  \"maximum_d\":" << maximum_d << ",\n"
            << "  \"minimum_endpoint_imbalance_slack\":" << minimum_endpoint_slack << ",\n"
            << "  \"minimum_power_bridge_slack\":" << minimum_power_slack << ",\n"
            << "  \"minimum_canonical_power_slack\":" << minimum_canonical_slack << ",\n"
            << "  \"minimum_canonical_prefix_slack\":"
            << minimum_canonical_prefix_slack << ",\n"
            << "  \"first_endpoint_imbalance_failure\":";
        write_optional(out, endpoint_failure);
        out << ",\n  \"first_power_bridge_failure\":";
        write_optional(out, power_failure);
        out << ",\n  \"first_canonical_power_failure\":";
        write_optional(out, canonical_failure);
        out << ",\n  \"first_canonical_prefix_failure\":";
        write_optional(out, canonical_prefix_failure);
        out << ",\n  \"minimum_endpoint_detail\":" << minimum_endpoint_detail
            << ",\n  \"minimum_power_detail\":" << minimum_power_detail
            << ",\n  \"minimum_canonical_detail\":" << minimum_canonical_detail
            << ",\n  \"minimum_canonical_prefix_detail\":"
            << minimum_canonical_prefix_detail
            << ",\n  \"digests\":{\"algorithm\":\"FNV-1a-64 little-endian\"," 
            << "\"classification_2_through_limit\":\"" << hex_u64(classification_digest)
            << "\",\"hard_h_d_s\":\"" << hex_u64(metric_digest) << "\"}\n}\n";
        if (!out) throw std::runtime_error("could not write output JSON");

        const auto seconds = std::chrono::duration<double>(
            std::chrono::steady_clock::now() - started
        ).count();
        std::cout << "limit=" << limit << " hard=" << hard_count
                  << " max_d=" << maximum_d
                  << " min_endpoint_slack=" << minimum_endpoint_slack
                  << " min_power_slack=" << minimum_power_slack
                  << " min_canonical_slack=" << minimum_canonical_slack
                  << " min_prefix_slack=" << minimum_canonical_prefix_slack
                  << " seconds=" << std::fixed << std::setprecision(3) << seconds << '\n';
        return 0;
    } catch (const std::exception& error) {
        std::cerr << "error: " << error.what() << '\n';
        return 1;
    }
}

File: problems/424/compute/wave5/C116_endpoint_invariant.py
#!/usr/bin/env python3
"""Exact prefix test for the C116 structural canonical-blocker invariants."""

from __future__ import annotations

import argparse
import hashlib
import json
import math
from pathlib import Path


OTHER = 0
GENERATED = 1
SPLITLESS = 2
HARD = 3
FNV_OFFSET = 14695981039346656037
FNV_PRIME = 1099511628211
MASK64 = (1 << 64) - 1


def require(condition: bool, message: object) -> None:
    if not condition:
        raise RuntimeError(message)


def allowed(n: int) -> bool:
    return n >= 2 and n % 3 != 1


def smallest_prime_factors(limit: int) -> list[int]:
    spf = list(range(limit + 1))
    for p in range(2, math.isqrt(limit - 1) + 1):
        if spf[p] != p:
            continue
        for multiple in range(p * p, limit, p):
            if spf[multiple] == multiple:
                spf[multiple] = p
    return spf


def divisors(n: int, spf: list[int]) -> list[int]:
    values = [1]
    while n > 1:
        prime = spf[n]
        old = tuple(values)
        power = 1
        while n % prime == 0:
            n //= prime
            power *= prime
            values.extend(value * power for value in old)
    return values


def admissible_pairs(n: int, spf: list[int]) -> list[tuple[int, int]]:
    product = n + 1
    pairs = []
    for left in divisors(product, spf):
        if left < 2 or left * left >= product:
            continue
        right = product // left
        if allowed(left) and allowed(right):
            pairs.append((left, right))
    return sorted(pairs)


def seed_root(endpoint: int) -> int:
    require(endpoint > 1 and endpoint % 2 == 1, ("odd endpoint", endpoint))
    shifted = endpoint - 1
    return 1 + shifted // (shifted & -shifted)


def fnv_byte(digest: int, value: int) -> int:
    return ((digest ^ value) * FNV_PRIME) & MASK64


def fnv_u64(digest: int, value: int) -> int:
    for shift in range(0, 64, 8):
        digest = fnv_byte(digest, (value >> shift) & 0xFF)
    return digest


def state_name(value: int) -> str:
    return ("other_hole", "generated", "structural_splitless", "hard")[value]


def certificate(
    h: int,
    pairs: list[tuple[int, int]],
    state: bytearray,
    spf: list[int],
    d: int,
    s: int,
    t: int,
    e0: set[int],
    e1: set[int],
) -> dict[str, object]:
    rows = []
    prefix_t = 0
    minimum_prefix_slack = None
    for index, (left, right) in enumerate(pairs, start=1):
        endpoints = []
        counted = False
        for endpoint in (left, right):
            present = state[endpoint] == GENERATED
            root = None if present else seed_root(endpoint)
            structural = False if present else not admissible_pairs(root, spf)
            counted = counted or structural
            endpoints.append(
                {
                    "value": endpoint,
                    "state": state_name(state[endpoint]),
                    "root": root,
                    "root_state": None if root is None else state_name(state[root]),
                    "root_pair_count": None if root is None else len(admissible_pairs(root, spf)),
                }
            )
        canonical = left if state[left] != GENERATED else right
        canonical_root = seed_root(canonical)
        canonical_structural = not admissible_pairs(canonical_root, spf)
        prefix_t += int(canonical_structural)
        prefix_slack = 2 * prefix_t - index + 8
        if minimum_prefix_slack is None or prefix_slack < minimum_prefix_slack:
            minimum_prefix_slack = prefix_slack
        rows.append(
            {
                "prefix_index": index,
                "prefix_t": prefix_t,
                "prefix_slack": prefix_slack,
                "pair": [left, right],
                "counted_in_s": counted,
                "canonical_blocker": canonical,
                "canonical_root": canonical_root,
                "counted_in_t": canonical_structural,
                "endpoints": endpoints,
            }
        )
    return {
        "h": h,
        "product": h + 1,
        "d": d,
        "s": s,
        "t": t,
        "E0_count": len(e0),
        "E1_count": len(e1),
        "endpoint_imbalance_slack": len(e0) + 8 - len(e1),
        "power_bridge_slack": 3 * s - d + 8,
        "canonical_power_slack": 2 * t - d + 8,
        "minimum_canonical_prefix_slack": minimum_prefix_slack,
        "E0": sorted(e0),
        "E1": sorted(e1),
        "pairs": rows,
    }


def scan(limit: int) -> dict[str, object]:
    require(limit >= 534, ("limit", limit))
    spf = smallest_prime_factors(limit + 2)
    state = bytearray(limit + 1)
    root_is_structural: dict[int, bool] = {}
    classification_digest = FNV_OFFSET
    metric_digest = FNV_OFFSET
    hard_count = 0
    maximum_d = 0
    first_failures: dict[str, dict[str, object] | None] = {
        "nonstructural_pairs_le_E1": None,
        "E0_le_twice_s": None,
        "endpoint_imbalance": None,
        "power_bridge": None,
        "canonical_t_le_s": None,
        "canonical_power": None,
        "canonical_prefix_balance": None,
    }
    minimum_endpoint_slack: tuple[int, dict[str, object]] | None = None
    minimum_power_slack: tuple[int, dict[str, object]] | None = None
    minimum_canonical_slack: tuple[int, dict[str, object]] | None = None
    minimum_canonical_prefix_slack: tuple[int, dict[str, object]] | None = None

    for n in range(2, limit + 1):
        pairs: list[tuple[int, int]] = []
        current = OTHER
        if n in (2, 3):
            current = GENERATED
        elif allowed(n):
            pairs = admissible_pairs(n, spf)
            if any(state[left] == GENERATED and state[right] == GENERATED for left, right in pairs):
                current = GENERATED
            elif not pairs:
                current = SPLITLESS
            elif n % 2 == 0:
                product = n + 1
                seed_three_easy = (
                    product % 3 == 0
                    and product // 3 != 3
                    and allowed(product // 3)
                )
                if not seed_three_easy:
                    current = HARD
        state[n] = current
        classification_digest = fnv_byte(classification_digest, current)
        if current != HARD:
            continue

        hard_count += 1
        d = len(pairs)
        maximum_d = max(maximum_d, d)
        e0: set[int] = set()
        e1: set[int] = set()
        s = 0
        t = 0
        prefix_t = 0
        hole_minimum_prefix_slack = None
        for index, (left, right) in enumerate(pairs, start=1):
            missing = 0
            structural_pair = False
            for endpoint in (left, right):
                if state[endpoint] == GENERATED:
                    continue
                missing += 1
                root = seed_root(endpoint)
                require(root < endpoint and root % 2 == 0 and allowed(root), (n, endpoint, root))
                require(state[root] != GENERATED, ("generated root", n, endpoint, root))
                if root not in root_is_structural:
                    root_is_structural[root] = not admissible_pairs(root, spf)
                structural = root_is_structural[root]
                require(structural == (state[root] == SPLITLESS), ("root state", n, endpoint, root))
                if structural:
                    e0.add(endpoint)
                    structural_pair = True
                else:
                    e1.add(endpoint)
            require(missing >= 1, ("unblocked hard pair", n, left, right))
            s += int(structural_pair)
            canonical = left if state[left] != GENERATED else right
            canonical_root = seed_root(canonical)
            canonical_structural = int(root_is_structural[canonical_root])
            t += canonical_structural
            prefix_t += canonical_structural
            prefix_slack = 2 * prefix_t - index + 8
            if (
                hole_minimum_prefix_slack is None
                or prefix_slack < hole_minimum_prefix_slack
            ):
                hole_minimum_prefix_slack = prefix_slack

        require(len(e0 & e1) == 0, ("endpoint partition", n))
        require(len(e0) + len(e1) >= d, ("missing endpoint count", n))
        metric_digest = fnv_u64(metric_digest, n)
        metric_digest = fnv_u64(metric_digest, d)
        metric_digest = fnv_u64(metric_digest, s)
        values = {
            "nonstructural_pairs_le_E1": (d - s, len(e1)),
            "E0_le_twice_s": (len(e0), 2 * s),
            "endpoint_imbalance": (len(e1), len(e0) + 8),
            "power_bridge": (d - 8, 3 * s),
            "canonical_t_le_s": (t, s),
            "canonical_power": (d - 8, 2 * t),
        }
        failed = [name for name, (lhs, rhs) in values.items() if lhs > rhs]
        require(hole_minimum_prefix_slack is not None, ("empty hard pair set", n))
        if hole_minimum_prefix_slack < 0:
            failed.append("canonical_prefix_balance")
        endpoint_slack = len(e0) + 8 - len(e1)
        power_slack = 3 * s - d + 8
        canonical_slack = 2 * t - d + 8
        need_row = (
            failed
            or minimum_endpoint_slack is None
            or endpoint_slack < minimum_endpoint_slack[0]
            or minimum_power_slack is None
            or power_slack < minimum_power_slack[0]
            or minimum_canonical_slack is None
            or canonical_slack < minimum_canonical_slack[0]
            or minimum_canonical_prefix_slack is None
            or hole_minimum_prefix_slack < minimum_canonical_prefix_slack[0]
        )
        row = certificate(n, pairs, state, spf, d, s, t, e0, e1) if need_row else {}
        for name in failed:
            if first_failures[name] is None:
                first_failures[name] = row
        if minimum_endpoint_slack is None or endpoint_slack < minimum_endpoint_slack[0]:
            minimum_endpoint_slack = (endpoint_slack, row)
        if minimum_power_slack is None or power_slack < minimum_power_slack[0]:
            minimum_power_slack = (power_slack, row)
        if minimum_canonical_slack is None or canonical_slack < minimum_canonical_slack[0]:
            minimum_canonical_slack = (canonical_slack, row)
        if (
            minimum_canonical_prefix_slack is None
            or hole_minimum_prefix_slack < minimum_canonical_prefix_slack[0]
        ):
            minimum_canonical_prefix_slack = (hole_minimum_prefix_slack, row)

    require(
        minimum_endpoint_slack is not None
        and minimum_power_slack is not None
        and minimum_canonical_slack is not None
        and minimum_canonical_prefix_slack is not None,
        "no hard holes",
    )
    return {
        "schema": "C116-endpoint-invariant-v2",
        "limit": limit,
        "exactness": {
            "arithmetic": "integers only",
            "closure": "independent ascending least-fixed-point reconstruction",
            "factor_pairs": "full SPF divisor enumeration",
            "structural_test": "literal empty admissible-pair set, cross-checked against state",
            "floating_point_acceptance": False,
        },
        "totals": {"hard_holes": hard_count, "maximum_d": maximum_d},
        "first_failures": first_failures,
        "minimum_endpoint_imbalance_slack": minimum_endpoint_slack[1],
        "minimum_power_bridge_slack": minimum_power_slack[1],
        "minimum_canonical_power_slack": minimum_canonical_slack[1],
        "minimum_canonical_prefix_slack": minimum_canonical_prefix_slack[1],
        "digests": {
            "algorithm": "FNV-1a-64 little-endian",
            "classification_2_through_limit": f"{classification_digest:016x}",
            "hard_h_d_s": f"{metric_digest:016x}",
        },
    }


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--limit", type=int, required=True)
    parser.add_argument("--output", type=Path, required=True)
    args = parser.parse_args()
    result = scan(args.limit)
    encoded = json.dumps(result, indent=2, sort_keys=True) + "\n"
    args.output.write_bytes(encoded.encode("ascii"))
    print(hashlib.sha256(encoded.encode("ascii")).hexdigest().upper())
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
